Vector 2.0 documentation changes

parnmatt opened the pull request.

Hunterness committed:
Start on relationship vector index documentation

There is at least these still left to fix:
- confirm version
- update the relationship example to an actual example
- add info about setting vector properties on relationships (kernel)
- add info about querying relationship vector indexes (kernel)
- more info regarding the two different providers? (kernel?)

parnmatt committed:
add queryRelationships and update signatures

parnmatt committed:
do not use 'deprecated'

parnmatt committed:
update signatures in vector-indexes

parnmatt committed:
add section of vector-1.0 differences

parnmatt committed:
add footnote xref for differences

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

Hunterness commented on modules/ROOT/pages/indexes/syntax.adoc:
```diff
@@ -198,7 +198,18 @@ ON (n.propertyName)
+----
+
+Vector indexes have two settings, `vector.dimensions` and `vector.similarity_function`, which have no default values.
+As of Neo4j 5.18, they have two index providers available, `vector-2.0` (default, prefered) and `vector-1.0` (compatibility).
```
```suggestion
As of Neo4j 5.18, they have two index providers available, `vector-2.0` (default, preferred) and `vector-1.0` (compatibility).
```

Hunterness commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -177,25 +194,63 @@ OPTIONS {indexConfig: {
+.Create a relationship vector index
+====
+
+// TODO: fix a proper example
```
This is still a relevant TODO

Hunterness commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -319,21 +374,21 @@ The index name can also be given as a parameter, `DROP INDEX $name`.
 [source,syntax]
 ----
-db.create.setNodeVectorProperty(node :: NODE, key :: STRING, vector :: LIST<FLOAT>)
+db.create.setNodeVectorProperty(node :: NODE, key :: STRING, vector :: LIST<INTEGER | FLOAT>)
```
I guess we still need to document the relationship set property procedure?

Hunterness replied:
Do we want an example of using it as well or are we happy with just listing it? (might be a question for docs team)

Hunterness commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -209,12 +264,12 @@ You can query a vector index using the procedure {link-procedures-reference}#pro
 [source,syntax,role="noheader",indent=0]
 ----
-db.index.vector.queryNodes(indexName :: STRING, numberOfNearestNeighbours :: INTEGER, query :: LIST<FLOAT>) :: (node :: NODE, score :: FLOAT)
+db.index.vector.queryNodes(indexName :: STRING, numberOfNearestNeighbours :: INTEGER, query :: LIST<INTEGER | FLOAT>) :: (node :: NODE, score :: FLOAT)
```
I guess we still need to document the query relationship index procedure? (in this file as well as it was only added to the syntax section)

Hunterness replied:
Do we want an example of using it as well or are we happy with just listing it? (might be a question for docs team)

parnmatt replied:
working on it :sweat_smile: just pushed it so docs can have a look

Hunterness reviewed (commented):
One minor typo and then three comments on things still needing to be fixed, is there anything else we're missing?

- [x] Once proper examples have been made, make a separate PR to the cypher cheat sheet with such examples as well (following what exists there for vector node indexes already) -> https://github.com/neo4j/docs-cheat-sheet/pull/158

parnmatt committed:
fix typo

parnmatt committed:
add setRelationshipVectorProperty

parnmatt committed:
add queryRelationships

parnmatt committed:
fix table

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

Hunterness commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -371,13 +387,13 @@ The index name can also be given as a parameter, `DROP INDEX $name`.
+As of Neo4j 5.13, you can set a vector property on a node using the {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`] procedure.
 It validates the input and sets the property as an array of {ieee-754} single precision values.
-This beta procedure replaces {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`].
+This beta procedure replaces {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`]. As of Neo4j 5.18, you can set a vector property on a relationship using the {link-procedures-reference}#procedure_db_create_setRelationshipVectorProperty[`db.create.setRelationshipVectorProperty`] procedure.
```
```suggestion
This beta procedure replaces {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`].
As of Neo4j 5.18, you can set a vector property on a relationship using the {link-procedures-reference}#procedure_db_create_setRelationshipVectorProperty[`db.create.setRelationshipVectorProperty`] procedure.
```
I assume docs will have a linebreak in between the sentences

Hunterness replied:
also is the `db.create.setNodeVectorProperty` still a beta procedure or should it be
```suggestion
This procedure replaces {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`].
As of Neo4j 5.18, you can set a vector property on a relationship using the {link-procedures-reference}#procedure_db_create_setRelationshipVectorProperty[`db.create.setRelationshipVectorProperty`] procedure.
```

parnmatt replied:
all of these `set` procedures are kept as "beta"

parnmatt committed:
add simple example of relationship vector index based on fulltext

parnmatt committed:
add linebreaks

parnmatt committed:
add setRelationshipVectorProperty example

parnmatt committed:
match signature output to use ANY

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

parnmatt committed:
add missing linebreak

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

parnmatt committed:
fix xref to GenAI

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -44,20 +46,30 @@ The procedures and commands for vector indexes are listed in the following table
+| Create node vector index.
 | `+CREATE VECTOR INDEX ...+`
-| label:new[Introduced in 5.15] Create a vector index for the specified label and property with the given vector dimensionality using the given similarity function.
+| label:new[Introduced in 5.15] Create a node vector index for the specified label and property with the given vector dimensionality using the given similarity function.
 See the xref:indexes/search-performance-indexes/managing-indexes.adoc#create-indexes[`CREATE INDEX`] command for more details. Replaces {link-procedures-reference}#procedure_db_index_vector_createNodeIndex[`db.index.vector.createNodeIndex`].
 
+| Create relationship vector index.
+| `+CREATE VECTOR INDEX ...+`
+| label:new[Introduced in 5.18] Create a relationship vector index for the specified relationship type and property with the given vector dimensionality using the given similarity function.
+See the xref:indexes/search-performance-indexes/managing-indexes.adoc#create-indexes[`CREATE INDEX`] command for more details.
```
Labels in a table should go at the end - I have an open PR that fixes this for the rest of the manual.
```suggestion
| Create a node vector index for the specified label and property with the given vector dimensionality using the given similarity function.
See the xref:indexes/search-performance-indexes/managing-indexes.adoc#create-indexes[`CREATE INDEX`] command for more details. Replaces {link-procedures-reference}#procedure_db_index_vector_createNodeIndex[`db.index.vector.createNodeIndex`].  label:new[Introduced in 5.15]

| Create relationship vector index.
| `+CREATE VECTOR INDEX ...+`
| Create a relationship vector index for the specified relationship type and property with the given vector dimensionality using the given similarity function.
See the xref:indexes/search-performance-indexes/managing-indexes.adoc#create-indexes[`CREATE INDEX`] command for more details. label:new[Introduced in 5.18]
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -89,8 +106,8 @@ Creating a vector index requires link:{neo4j-docs-base-uri}/operations-manual/{p
 
-A vector index is a single-label, single-property index for nodes.
-A vector index needs to be configured with both the dimensionality of the vector (`INTEGER` between `1` and `2048` _inclusive_), and the measure of similarity between two vectors (case-insensitive `STRING`).
+A vector index is a single-label, single-property index for nodes or a single-relationship-type, single-property index for relationships.{fn-provider-differences}
+A vector index needs to be configured with both the dimensionality of the vector (`INTEGER` between `1` and `4096` _inclusive_),{fn-provider-differences} and the measure of similarity between two vectors (case-insensitive `STRING`).
```
Two references to the same footnote, without another footnote in between, in the same paragraph might not be necessary. 

```suggestion
A vector index is a single-label, single-property index for nodes or a single-relationship-type, single-property index for relationships.
A vector index needs to be configured with both the dimensionality of the vector (`INTEGER` between `1` and `4096` _inclusive_), and the measure of similarity between two vectors (case-insensitive `STRING`).{fn-provider-differences}
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -109,6 +126,16 @@ OPTIONS "{" option: value[, ...] "}"
+ON (r.propertyName)
+OPTIONS "{" option: value[, ...] "}"
+----
+| Create a vector index on relationships.  label:new[Introduced in 5.18]
+The options map is mandatory because setting the vector dimensions and similarity function is mandatory when creating a vector index.
```
```suggestion
| Create a vector index on relationships. 
The options map is mandatory because setting the vector dimensions and similarity function is mandatory when creating a vector index. label:new[Introduced in 5.18]
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -122,7 +149,7 @@ The index name must be unique among all indexes and constraints.
 
 The `CREATE VECTOR INDEX` command take an `OPTIONS` clause. This has two parts, the `indexProvider` and `indexConfig`.
-The provider can only have the default value, `'vector-1.0'`.
+As of Neo4j 5.18, there are two available index providers, `vector-2.0` (default, preferred) and `vector-1.0` (compatibility).
```
I'd be inclined to just go with "(default)" here (it kind of assumes that it is preferred), and also remove "(compatibility)" which in it self isnt really telling the user anything? Or am I missing something?

```suggestion
As of Neo4j 5.18, there are two available index providers, `vector-2.0` (default) and `vector-1.0`.
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -283,7 +352,54 @@ The results are expected, with papers discussing graph-based nearest-neighbor se
 
+.Query a relationship vector index
+====
+This example takes a query vector describing the themes and and nuances we are looking for in the reviews.
```
```suggestion
This example takes a query vector describing particular themes and nuances in the reviews.
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -345,6 +468,15 @@ MATCH (n:Node {id: $id})
 CALL db.create.setNodeVectorProperty(n, 'propertyKey', $vector)
 RETURN n
 ----
+Likewise by matching a relationship and setting its vector properties using `db.create.setRelationshipVectorProperty`:
```
```suggestion
Likewise, by matching a relationship and setting its vector properties using `db.create.setRelationshipVectorProperty`:
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -424,6 +557,25 @@ Cosine similarity rebounds the trigonometric cosine.
 In the above equation the trigonometric cosine is given by the scalar product of the two unit vectors.
 
+[[indexes-vector-provider-differences]]
+== Compatibility index provider differences
```
```suggestion
== Vector index providers and compatibility

```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -424,6 +557,25 @@ Cosine similarity rebounds the trigonometric cosine.
 
+[[indexes-vector-provider-differences]]
+== Compatibility index provider differences
+The default and preferred vector index provider is `vector-2.0`.
```
Can we mention something brief about the differences between the 2 providers? What has improved with 2.0?

```suggestion
As of Neo4j 5.18, the default and preferred vector index provider is `vector-2.0`.
```

JPryce-Aklundh commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -424,6 +557,25 @@ Cosine similarity rebounds the trigonometric cosine.
+Previously created `vector-1.0` indexes will continue to function.
+New indexes can still be created with the `vector-1.0` index provider if specified, see xref:indexes-vector-create[].
+
+Supported index schema::
+A single-label, single-property index for nodes.
+
+Supported property value type::
+`LIST<FLOAT>`.
+
+Supported vector dimensionality::
+`INTEGER` between `1` and `2048` _inclusive_.
+
+Cosine similarity vector validity::
+* All vector components can be represented finitely in {ieee-754} single precision.
+* Its {l2-norm} is non-zero and can be represented finitely in {ieee-754} single precision.
```
What does this table have to do with the differences between vector-1.0 and vector-2.0? Or should it have its own heading?

Hunterness replied:
If I have understood it correctly, those are the things `vector-1.0` follows that differs from what `vector-2.0` does (which is in the main parts with footnotes to this section)

JPryce-Aklundh commented on modules/ROOT/pages/indexes/syntax.adoc:
```diff
@@ -198,7 +198,18 @@ ON (n.propertyName)
+----
+
+Vector indexes have two settings, `vector.dimensions` and `vector.similarity_function`, which have no default values.
+As of Neo4j 5.18, they have two index providers available, `vector-2.0` (default, preferred) and `vector-1.0` (compatibility).
```
```suggestion
As of Neo4j 5.18, they have two index providers available, `vector-2.0` (default) and `vector-1.0`.
```

Hunterness commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -177,25 +204,61 @@ OPTIONS {indexConfig: {
+(:Manager)-[:REVIEWED]->(:Employee)
+----
+
+Assume for each abstract, you have generated a 256-dimensional vector `embedding` of the review's `text` using a shortening of Open AI's `text-embedding-3-large` model.
```
```suggestion
Assume for each review, you have generated a 256-dimensional vector `embedding` of the review's `text` using a shortening of Open AI's `text-embedding-3-large` model.
```

JPryce-Aklundh reviewed (commented):
Thanks @parnmatt! Looks good overall - my main question was about the new section on provider compatibility (particularly the bullet point list in it, which did not fit to me).

Also, and this goes to @Hunterness comments as well, I think we should start thinking about how to include a "proper" example for this page, which users can replicate and query along with for themselves. This will not be possible for this PR, but the lack of it on this page (which is quite popular) is the source of continued feedback to our dashboard.

Hunterness commented on modules/ROOT/pages/indexes/semantic-indexes/vector-indexes.adoc:
```diff
@@ -451,7 +601,17 @@ The following table lists the known issues and the version in which they were fi
 | Known issues | Fixed in
 
-| Vector indexes cannot be assigned autogenerated names. 
+| Procedure signatures from `SHOW PROCEDURES` will render the vector arguments with a type of `ANY` rather than the semantically correct type of `LIST<INTEGER \| FLOAT>`.
+[NOTE]
+====
+The types are still enforced as `LIST<INTEGER \| FLOAT>`.
+====
+|
```
if it isn't fixed should it be up in the `Limitations and idiosyncrasies` instead of this table (from what I've seen this holds the fixed issues and that the open ones?)

JPryce-Aklundh replied:
We will keep it here for now, as it still might be fixed - but will change to your suggestion if fix becomes less probable

parnmatt committed:
make a table of differences

though I would prefer to use `a` column, they don't play well with
footnotes (and the suggestion of a new column mode _still_ hasn't been
accepted into the language).

Thus "hacking" paragraphs, with a non breaking space, to give larger
separation to improve readability.

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

JPryce-Aklundh committed:
Merge branch 'dev' into 5.18-vector-2.0

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

JPryce-Aklundh committed:
editorial fix

neo-technology-commit-status-publisher commented:
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-920.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Create, show, and delete indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/search-performance-indexes/managing-indexes)
- [Vector search indexes](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/semantic-indexes/vector-indexes)
- [Syntax](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/indexes/syntax)
- [Operators in detail](https://neo4j-docs-cypher-920.surge.sh/cypher-manual/5/planning-and-tuning/operators/operators-detail)
<!-- Sticky Pull Request Commentdocs-pr-changes -->

JPryce-Aklundh committed:
release information and label fixes

neo-technology-commit-status-publisher commented:
Thanks for the documentation updates.

The preview documentation has now been torn down - reopening this PR will republish it.
<!-- Sticky Pull Request Commentdocs-pr-changes -->
